sbi/inference/trainers/marginal/marginal_base.py

```diff
@@ -0,0 +1,392 @@
+# This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
+# under the Apache License Version 2.0, see <https://www.apache.org/licenses/>
+
+import time
+from copy import deepcopy
+from datetime import datetime
+from pathlib import Path
+from typing import Callable, Dict, Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+from torch.nn.utils.clip_grad import clip_grad_norm_
+from torch.optim.adam import Adam
+from torch.utils import data
+from torch.utils.data.sampler import SubsetRandomSampler
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from sbi.neural_nets.estimators import UnconditionalDensityEstimator
+from sbi.neural_nets.estimators.shape_handling import (
+    reshape_to_batch_event,
+)
+from sbi.neural_nets.factory import marginal_nn
+from sbi.utils import check_estimator_arg, get_log_root
+from sbi.utils.torchutils import assert_all_finite, process_device
+
+
+class MarginalTrainer:
+    def __init__(
+        self,
+        density_estimator: Union[str, Callable] = "MAF",
+        device: str = "cpu",
```

Comment:
can you add a test for gpu as well?

Reply:
I'm not sure of what you mean with this. Do you have a similar example please?